File: strongPythonStrongerAI/task_prioritizer.py
import random

class Task():
    def __init__(self, name, deadline):
        self.name = name
        self.deadline = deadline

t1 = Task(name="Do physics assignment",deadline=6)
# print(f"Task: {t1.name}\tDeadline: {t1.deadline}")
tasks = []

def SORT (taskList: list):
    # for(i=0; i<len(taskList)-1; i++):
    for i in range(len(taskList)-1):
        for j in range(len(taskList)-i-1):
            if(taskList[j].deadline > taskList[j+1].deadline):
                taskList[j], taskList[j+1] = taskList[j+1], taskList[j]
    return taskList

def MPT(tasks: list):
    result = SORT(tasks)
    print("Most prior task: ")
    print(f"Task: {result[0].name}  |   Deadline: {result[0].deadline}")

def LPT(tasks: list):
    result = SORT(tasks)
    print("Least prior task: ")
    print(f"Task: {result[len(tasks)-1].name}  |   Deadline: {result[len(tasks)-1].deadline}")

def ART(tasks: list):
    result = random.choice(tasks)
    print("Randomly Assigned task: ")
    print(f"Task: {result.name}  |   Deadline: {result.deadline}")
        

while True:
    
    while True:
        ch = input("Do you want to enter task (y/n): ")
        if ch.lower() == 'n':
            break
        t = Task(input("Enter task: "),int(input("Enter deadline: ")))
        tasks.append(t)
    
    print("""Enter following number to perform operation:
          1: Most prior task
          2: Least prior task
          3: Assign random task
          0: Exit !""")
    num = int(input("Enter your choice: "))

    match num:
        case 1:
            MPT(tasks)
        case 2:
            LPT(tasks)
        case 3:
            ART(tasks)
        case 0:
            break
        case _:
            print("Invalid number entered !\nTry again !!")

"""
concept:
I made user-defined functions like MPT, LPT, ART, SORT, but these could be done in few line of codes and do the work.

most_prior = min(tasks, key=lambda t: t.deadline)   # smallest deadline | MPT
least_prior = max(tasks, key=lambda t: t.deadline)  # largest deadline | LPT
and no need to sort.
and if i had to sort, it could have been done by

sorted(taskList, key=lambda t: t.deadline)

I did it the long way because, i didn't know inbuild functions like these exist in python.

"""

File: strongPythonStrongerAI/task_prioritizer.cpp
#include <iostream>
#include<string>
#include <vector>
#include <algorithm>
using namespace std;

// Structure to store task details
struct Task {
    string name;
    int deadline; // in days
};

// Function to sort tasks based on priority (lower deadline = higher priority)
void sortTasksByPriority(vector<Task> &tasks) {
    sort(tasks.begin(), tasks.end(), [](const Task &a, const Task &b) {
        return a.deadline < b.deadline;
    });
}

// Function to get the highest priority task (smallest deadline)
Task getHighestPriorityTask(const vector<Task> &tasks) {
    if (tasks.empty()) {
        return {"No Task", -1}; // return dummy task if empty
    }
    return tasks[0]; // after sorting, first task is highest priority
}

int main() {
    vector<Task> tasks;
    char choice;

    while (true) {
        Task t;
        cout << "Enter task name: ";
        cin.ignore(); // clear buffer before getline
        getline(cin, t.name);

        cout << "Enter deadline in days: ";
        cin >> t.deadline;

        tasks.push_back(t);

        cout << "Do you want to add another task? (y/n): ";
        cin >> choice;
        if (choice == 'n' || choice == 'N') {
            break;
        }
    }

    // Sort tasks by priority
    sortTasksByPriority(tasks);

    cout << "\nTasks sorted by priority:\n";
    for (const auto &t : tasks) {
        cout << "Task: " << t.name << " | Deadline: " << t.deadline << " days\n";
    }

    // Get the highest priority task
    Task topTask = getHighestPriorityTask(tasks);
    cout << "\nHighest Priority Task:\n";
    cout << "Task: " << topTask.name << " | Deadline: " << topTask.deadline << " days\n";

    return 0;
}
